python/cuml/ensemble/randomforest.pyx

```diff
@@ -36,81 +36,160 @@ from cuml.utils import get_cudf_column_ptr, get_dev_array_ptr, \
+                          float *, int *,
+                          int, int, int *, bool) except +
+
     cdef RF_metrics score(cumlHandle& handle,
-                          rfClassifier[double] *, double *, int *,
-                          int, int, int *, bool)
+                          RandomForestMetaData[double, int] *,
+                          double *, int *,
+                          int, int, int *, bool) except +
+
+    cdef void print_rf_summary(RandomForestMetaData[float, int] *) except +
+    cdef void print_rf_summary(RandomForestMetaData[double, int] *) except +
+
+    cdef void print_rf_detailed(RandomForestMetaData[float, int] *) except +
+    cdef void print_rf_detailed(RandomForestMetaData[double, int] *) except +
+
+    # cdef void print(RF_params) except +
 
     cdef RF_params set_rf_class_obj(int, int, float,
                                     int, int, int,
                                     bool, bool, int, float, CRITERION,
                                     bool) except +
 
+    # Random Forest Regressor
+    cdef void fit(cumlHandle & handle,
+                  RandomForestMetaData[float, float] *,
+                  float *,
+                  int,
+                  int,
+                  float *,
+                  RF_params) except +
+
+    cdef void fit(cumlHandle & handle,
+                  RandomForestMetaData[double, double] *,
+                  double *,
+                  int,
+                  int,
+                  double *,
+                  RF_params) except +
+
+    cdef void predict(cumlHandle& handle,
+                      RandomForestMetaData[float, float] *,
+                      float *,
+                      int,
+                      int,
+                      float *,
+                      bool) except +
+
+    cdef void predict(cumlHandle& handle,
+                      RandomForestMetaData[double, double] *,
+                      double *,
+                      int,
+                      int,
+                      double *,
+                      bool) except +
+
+    cdef RF_metrics score(cumlHandle& handle,
+                          RandomForestMetaData[float, float] *,
+                          float *, float *,
+                          int, int, float *, bool) except +
+
+    cdef RF_metrics score(cumlHandle& handle,
+                          RandomForestMetaData[double, double] *,
+                          double *, double *,
+                          int, int, double *, bool) except +
+
+    cdef RF_params set_rf_class_obj(int, int, float,
+                                    int, int, int,
+                                    bool, bool, int, float, CRITERION,
+                                    bool) except +
 
 cdef class RandomForest_impl():
 
     cpdef object handle
-    cdef rfClassifier[float] *rf_classifier32
-    cdef rfClassifier[double] *rf_classifier64
+    cdef RandomForestMetaData[float, int] *rf_forest32
```

Comment:
We dont need a cython class in the wrapper if the model is stateless.

Reply:
`RandomForestMetaData` is a POD struct as per the guidelines here  https://github.com/rapidsai/cuml/blob/branch-0.9/cpp/DEVELOPER_GUIDE.md#stateless-c-api 
which will hold pointers to tree nodes, RF params info.

This is simply passed to the fit/predict functions and is not owned by our implementation, contrary to what was the case before. I don't see how I can avoid this or how this would make the model stateful. Could you please clarify?

Also tagging @teju85 for clarifications.

Reply:
First fast look (sorry for getting late to see this PRs, lots of big PRs to review lately), and it seems that indeed this POD struct works well! I think that then we can change the cython side to not have a cdef class similar to UMAP now with the struct being in the normal Python class like UMAP https://github.com/rapidsai/cuml/blob/8e8bd657a6f2bb832cd6df8ab7651564d1c89190/python/cuml/manifold/umap.pyx#L230

But that can be a follow up PR to change the python code and make it picklable and test and whatnot. I think this PR should focus on making the C++ side change with the minimal changes to python just to make it work as is

Reply:
Thanks to both for clarifying! Yes, that makes sense. And I absolutely agree with having this PR focus on the C++ side; that's the reason I made minimal Python wrapper changes.